core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -98,8 +99,23 @@ class ReplicaAlterLogDirsThread(name: String,
       throw new IllegalStateException("Offset mismatch for the future replica %s: fetched offset = %d, log end offset = %d.".format(
         topicPartition, fetchOffset, futureReplica.logEndOffset.messageOffset))
 
-    // Append the leader's messages to the log
-    partition.appendRecordsToFutureReplica(records)
+    try {
+      partition.appendRecordsToFutureReplica(records)
+    } catch {
+      case e: UnexpectedAppendOffsetException =>
+        if (e.firstOffset.isDefined && futureReplica.logEndOffset.messageOffset == futureReplica.logStartOffset) {
+          // This may happen if the log start offset on the current replica falls in the middle of the
+          // batch due to delete records request and the future replica tries to fetch its first offset
+          // from the current replica. We will truncate fully again, so that the log segment starts from
+          // the base offset that is the base offset of the batch, and try to append records again.
+          info(s"Unexpected offset in append to $topicPartition. First offset ${e.firstOffset.get} is less than log start offset ${futureReplica.logStartOffset}." +
+               s" Since this is the first record to be appended to the future replica's log, will start the log from offset ${e.firstOffset.get}.")
+          partition.truncateFullyAndStartAt(e.firstOffset.get, isFuture = true)
+          partition.appendRecordsToFutureReplica(records)
+        } else
+          throw e
+    }
```

Comment:
Can we not share this code with the other class? Also, instead of `isDefined` followed by `get`, we should use pattern matching.